### PR TITLE
plugins/lsp: add omnisharp language server

### PR DESCRIPTION
**Pull request by garaiza-93**
Adds option to enable omnisharp, and options to configure it.

### CODE REVIEW

**Comment by Tanish2002** [comment]
Haven't tried this pr myself yet. But I did manually once add OmniSharp to my config and If I remember OmniSharp is kinda broken for nvim-lsp (https://github.com/OmniSharp/omnisharp-roslyn/issues/2483)

If it's still true then I believe it's better to not merge this for now since it won't work anyways. We could potentially add csharp-ls instead. Though it's not packaged in nixpkgs

**Comment by garaiza-93** [comment]
> But I did manually once add OmniSharp to my config

When was this? And I'm assuming you installed omnisharp through nixpkgs?

**Comment by Tanish2002** [comment]
> When was this? And I'm assuming you installed omnisharp through nixpkgs?

Yeah I installed it through nixpkgs. You can check if produces some problem. I'll check it out tomorrow

**Comment by garaiza-93** [comment]
I have my fork if you want to use it and try it out. I
'm probably fortunate enough to have a neovim config with an older version of omnisharp that doesn't have to deal with this.

That issue you linked goes back to December 2022, and i'm not really holding my breath for a fix. maybe i'll bring in that hacky workaround lol

**Comment by garaiza-93** [comment]
I found some stuff that suggests that this won't be a problem for long.
 There's this open PR to bring omnisharp back in line with the LSP spec: https://github.com/OmniSharp/omnisharp-roslyn/pull/2520
 
 Then `nvim-lspconfig` itself seemed to add a workaround to fix this: https://github.com/neovim/nvim-lspconfig/pull/2692

**Comment by Tanish2002** [comment]
If omnisharp is going to fix it then it's pretty nice. We can probably wait for them to merge the changes.
Also nvim-lspconfig's version is dependent on nixpkgs. If it's updated then we can simply update flake inputs.

Other than that I just tried out your fork and yes the problem is still there. I just created a new console project using `dotnet new console` and as soon as I opened the `Program.cs` file I got the errors
image

**Comment by garaiza-93** [comment]
Just checked and lspconfig's fix was only merged in 2 days ago, and `vimPlugins` hasn't been updated since the end of May.

:pensive: I think I'll leave this PR open until nixpkgs gets updated. I'm tempted to add an input just for nvim-lspconfig to point towards the most recent version, but I doubt that would be accepted lol

**Comment by GaetanLepage** [comment]
Yes I am on the team of waiting for a fix if it is likely to come.

File: plugins/lsp/language-servers/default.nix
Comment (Tanish2002) [inline plugins/lsp/language-servers/default.nix:383]: Not sure if if it's necessary or not to do this. @GaetanLepage can probably tell.
But we do have helper functions for these like `helpers.defaultNullOpts.mkBool false "some description";`
You can probably look into the `nixd` server declaration just above yours. (You can do this for all the options)

**Review by Tanish2002 — CHANGES_REQUESTED** [review]

**Comment by Tanish2002** [comment]
> 😔 I think I'll leave this PR open until nixpkgs gets updated. I'm tempted to add an input just for nvim-lspconfig to point towards the most recent version, but I doubt that would be accepted lol

Until then you can probably look into the changes I asked for! 😃

**Comment by Tanish2002** [comment]
> Yes I am on the team of waiting for a fix if it is likely to come.

The omnisharp team hasn't fixed that issue for months. Since nvim-lspconfig has added a workaround into the plugin themselves we can just merge this PR when nixpkgs update nvim-lspconfig instead of waiting for omnisharp team's fix.

**Comment by garaiza-93** [comment]
Speaking of, lspconfig's fix is on the way.
Here's lspconfig currently: https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/editors/vim/plugins/generated.nix#L6866

https://github.com/NixOS/nixpkgs/pull/240130 has been merged and updates to a version of lspconfig that should include that fix. [It's this line specifically](https://github.com/NixOS/nixpkgs/pull/240130/commits/8263818e66553ba07933e753eeae72d6b1e2f440#diff-54bef7a6c9c5497e77086f7935fbfe5c45e285f36f12f15b7044f42430e8e3ebR6890)

So, in a couple of days, this should be good to go. Let's leave this PR open until then just in case.

**Comment by Tanish2002** [comment]
The OmniSharp fix was merged as well.

**Comment by GaetanLepage** [comment]
> The OmniSharp fix was merged as well.

Great, I just updated the `flake.lock` for nixvim, so it should be in it.
Can you rebase from `main` and re-test please ?

File: plugins/lsp/language-servers/default.nix
Comment (GaetanLepage) [reply plugins/lsp/language-servers/default.nix:383]: Indeed you are right @Tanish2002

**Comment by Tanish2002** [comment]
> Great, I just updated the `flake.lock` for nixvim, so it should be in it. Can you rebase from `main` and re-test please ?

The fix was just merged. nixpkgs will not have the fix until omnisharp creates a new release with the fix. (Correct me if I'm wrong)

**Comment by GaetanLepage** [comment]
> The fix was just merged. nixpkgs will not have the fix until omnisharp creates a new release with the fix. (Correct me if I'm wrong)

I thought the "fix" was in `nixpkgs`, not in `omnisharp` itself. What is [this PR](https://github.com/NixOS/nixpkgs/pull/240130) about then ?

**Comment by Tanish2002** [comment]
> I thought the "fix" was in `nixpkgs`, not in `omnisharp` itself. What is [this PR](https://github.com/NixOS/nixpkgs/pull/240130) about then ?

Ok so......

OmniSharp has had that issue for months without a fix. @garaiza-93 said nvim-lspconfig added a workaround to the plugin itself instead of waiting for OmniSharp to fix it.
I believed it at that time and didn't check the nvim-lspconfig's PR myself. I just checked it out and... The changes it does doesn't fix the issue I was referring to. That PR is fixing some other things.

Other than that OmniSharp themselves have finally fixed the issue themselves a few hours ago. Though a release hasn't been created yet. 

So basically we'll have to wait for OmniSharp to create a release and it getting into nixpkgs before merging 

Link to convo about both PRs: https://github.com/pta2002/nixvim/pull/465#issuecomment-1616090847

File: plugins/lsp/language-servers/default.nix
Comment (garaiza-93) [reply plugins/lsp/language-servers/default.nix:375]: I'm not sure if I understand. What should I change the command to?

**Comment by GaetanLepage** [comment]
Thanks @Tanish2002 for the clarification. I would advise to indeed wait for a release of omnisharp.

File: plugins/lsp/language-servers/default.nix
Comment (Tanish2002) [reply plugins/lsp/language-servers/default.nix:375]: just this `cmd = cfg: ["${cfg.package}/bin/OmniSharp"];` should be fine

**Comment by garaiza-93** [comment]
Alright, waiting for an updated release on omnisharp to reach nixpkgs.

File: plugins/lsp/language-servers/default.nix
Comment (Tanish2002) [inline plugins/lsp/language-servers/default.nix:383]: can you please add a new line after every new settingsOption.

**Review by Tanish2002 — CHANGES_REQUESTED** [review]

**Review by Tanish2002 — APPROVED** [review]

**Comment by GaetanLepage** [comment]
I made a PR on nixpkgs to update omnisharp: https://github.com/NixOS/nixpkgs/pull/244119
PR status: https://nixpk.gs/pr-tracker.html?pr=244119

**Comment by GaetanLepage** [comment]
Hey @garaiza-93 could you please rebase your branch on `main` please ?
It seems that it is a bit messed up right now.

**Review by GaetanLepage — APPROVED** [review]

**Pull request merged**
